File: sources/go/list/CircleList.go
/**
 * 循环链表
 */

package list

import (
	"fmt"
)

// 循环链表结点类型
type CircleNode struct {
	data	interface{}
	next 	*CircleNode
}

// 循环链表
type CircleList struct {
	head	*CircleNode
	size  int
}

// 构造Node
func newCircleNode(e interface{}) *CircleNode{
	return &CircleNode {
		data: e,
		next: nil,
	}
}

// 构造循环链表
func NewCircleList() *CircleList{
	p := newCircleNode(0)
	p.next = p 	// 循环链表！！！
	return &CircleList{
		head: p,
		size: 0,
	}
}

// 增：插入结点
// 约定：带头结点的链表，插入时，只能在头结点之后插入，也不允许插入超过最大元素个数的位置
func (l *CircleList)Insert(e interface{}, index int){
	if index < 1 || index > l.size + 1 {
		fmt.Println("插入位置不合法")
		return
	}

	// 找到插入位置前一个位置：也可以使用 Locate函数
	p := l.head
	k := 0
	for p.next != l.head && k < index - 1 {
		p = p.next
		k++
	}

	// 创建要插入的节点
	q := newCircleNode(e)
	q.next = p.next
	p.next = q

	l.size++
}

// 删：根据位置删除，返回被删除的元素
func (l *CircleList)Delete(index int) interface{}{ 
	if index < 1 || index > l.size {
		fmt.Println("删除位置非法")
		return nil
	}

	// 找到删除位置前一个位置：也可以使用 Locate函数
	p := l.head
	k := 0
	for p.next != l.head && k < index - 1 {
		p = p.next
		k++
	}

	// 执行删除
	q := p.next
	tempData := q.data
	p.next = q.next

	l.size--
	return tempData
}

// 改
func (l *CircleList)Update(index int, e interface{}){
	p := l.Locate(index)
	if p == nil {
		return
	}
	p.data = e
}

// 查
func (l *CircleList)Search(e interface{}) *CircleNode{
	p := l.head
	for p.next != l.head {
		if p.data == e {
			return p
		}
		p = p.next
	}
	return nil
}

// 定位
func (l *CircleList)Locate(index int) *CircleNode{
	if index < 0 || index > l.size + 1 {
		fmt.Println("获取位置不合法")
		return nil
	}

	p := l.head
	k := 0
	for p.next != l.head && k < index {
		p = p.next
		k++
	}
	return p
}

// 获取表长度
func (l *CircleList)Length() int {
	return l.size
}

// 清空表：仅保留头节点
func (l *CircleList)Clear() {
	l.head.next = l.head
	l.size = 0
}

// 显示循环链表
func (l *CircleList)Display(){
	if l.size == 0 {
		fmt.Println("空链表")
		return
	}

	p := l.head
	pos := 0
	for p.next != nil {
		if pos == l.size {
			fmt.Print(p.data, "->", l.head.data, "->...\n", )
			break
		}
		fmt.Print(p.data, "->")
		p = p.next
		pos++
	}
}

File: sources/js/list/CircleList.js
/**
 * 循环链表
 */

class Node {
    constructor(data) {
        this.data = data
        this.next = null
    }
}

class CircleList {
    constructor() {
        let p = new Node(0)
        p.next = p
        this.head = p
        this.size = 0
    }
    // 增：插入结点
    // 约定：带头结点的链表，插入时，只能在头结点之后插入，也不允许插入超过最大元素个数的位置
    insert(e, index) {
        if (index < 1 || index > this.size + 1) {
            console.log('插入位置不合法')
            return
        }

        // 找到插入位置前一个位置：也可以使用 Locate函数
        let p = this.head
        let k = 0
        while (p.next != this.head && k < index - 1) {
            p = p.next
            k++
        }

        // 创建要插入的节点
        let q = new Node(e)
        q.next = p.next
        p.next = q

        this.size++
    }
    // 删：根据位置删除，返回被删除的元素
    delete(index) {
        if (index < 1 || index > this.size) {
            console.log('删除位置非法')
            return
        }
        // 找到删除位置前一个位置：也可以使用 Locate函数
        let p = this.head
        let k = 0
        while (p.next != this.head && k < index - 1) {
            p = p.next
            k++
        }
        // 执行删除
        let q = p.next
        let tempData = q.data
        p.next = q.next

        this.size--
        return tempData
    }
    // 改
    update(index, e) {
        let p = this.locate(index)
        if (!p) {
            return
        }
        p.data = e
    }
    // 查
    search(e) {
        let p = this.head
        while (p.next != this.head) {
            if (p.data == e) {
                return p
            }
            p = p.next
        }
    }
    // 定位
    locate(index) {
        if (index < 0 || index > this.size + 1) {
            console.log('获取位置不合法')
            return
        }

        let p = this.head
        let k = 0
        while (p.next != this.head && k < index) {
            p = p.next
            k++
        }
        return p
    }
    // 获取长度
    length() {
        return this.size
    }
    // 清空表
    clear() {
        this.head.next = this.head
        this.size = 0
    }
    // 显示表
    display() {
        if (this.size == 0) {
            console.log('空链表')
            return
        }

        let p = this.head
        let pos = 0
        let res = ''
        while (p.next) {
            if (pos == this.size) {
                res += p.data
                res += '->'
                res += this.head.data
                res += '->...'
                break
            }
            res += p.data
            res += '->'
            p = p.next
            pos++
        }
        console.log(res)
    }
}

module.exports = CircleList
